dht_net.go

```diff
@@ -112,17 +117,43 @@ func (dht *IpfsDHT) updateFromMessage(ctx context.Context, p peer.ID, mes *pb.Me
 	return nil
 }
 
-func (dht *IpfsDHT) messageSenderForPeer(p peer.ID) *messageSender {
+func (dht *IpfsDHT) messageSenderForPeer(p peer.ID) (*messageSender, error) {
+	dht.smlk.Lock()
+	ms, ok := dht.strmap[p]
+	if ok {
+		dht.smlk.Unlock()
+		return ms, nil
+	}
+	dht.smlk.Unlock()
+
+	ms = &messageSender{p: p, dht: dht}
+	if err := ms.prep(); err != nil {
```

Comment:
Can't we just make a single `messageSender` and synchronize on its lock for `prep`?

Reply:
we cant do that without holding the dht.smlk for the duration of the prep call. And the prep call may take arbitrarily long (up to a dial timeout). We don't want to block every peer from calling this message waiting on a dial. :/

Reply:
We can sync on the lock outside the `smlk`.
I was thinking of something like this:
```
ms = &messageSender{...}
dht.strmap[p] = ms
dht.smlk.Unlock()

ms.lk.Lock()
if err := ms.prep(); ...
```

Reply:
We can even acquire the `ms.lk` inside `smlk`

Reply:
Obviously, the subsequent logic would have to be adjusted, but i think it might result in simpler code.

Reply:
Then what happens when two separate goroutines have the same message sender (one created it and one got it from the map and is waiting on the prep lock) and then prep fails? goroutine that called prep exits, and the goroutine that just got a reference to the messageSender and was waiting on the lock now tries to re-call prep, which is okay if prep fails again, but if prep succeeds on that second call for some reason, then we have an open stream attached to a message sender that isnt being tracked by anything.

Reply:
yeah, this is the problem. I had a similar issue in #92 and resolved it with a reset count, but it's hairy.